### PR TITLE
Add A128CBCHS256 support

### PR DESCRIPTION
**Pull request by ramunasjurgilas**

### CODE REVIEW

File: JOSESwift/Sources/Algorithms.swift [hunk 1]
```diff
@@ -54,32 +55,41 @@ public enum AsymmetricKeyAlgorithm: String, CaseIterable {
...
 public enum SymmetricKeyAlgorithm: String {
     case A256CBCHS512 = "A256CBC-HS512"
+    case A256CBCHS256 = "A256CBC-HS256"
 
     var hmacAlgorithm: HMACAlgorithm {
         switch self {
         case .A256CBCHS512:
             return .SHA512
+        case .A256CBCHS256:
+            return .SHA256
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: 👍

File: JOSESwift/Sources/Algorithms.swift [hunk 2]
```diff
@@ -54,32 +55,41 @@ public enum AsymmetricKeyAlgorithm: String, CaseIterable {
...
+            return 32
         }
     }
 
     var initializationVectorLength: Int {
         switch self {
         case .A256CBCHS512:
             return 16
+        case .A256CBCHS256:
+            return 16
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: 👍

File: JOSESwift/Sources/Algorithms.swift [hunk 3]
```diff
@@ -107,11 +125,14 @@ public enum SymmetricKeyAlgorithm: String {
...
 public enum HMACAlgorithm: String {
     case SHA512 = "SHA512"
+    case SHA256 = "SHA256"
 
     var outputLength: Int {
         switch self {
         case .SHA512:
             return 64
+        case .SHA256:
+            return Int(CC_SHA256_DIGEST_LENGTH)
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]:
```suggestion
            return 32
```

File: JOSESwift/Sources/Algorithms.swift [hunk 4]
```diff
@@ -54,32 +55,41 @@ public enum AsymmetricKeyAlgorithm: String, CaseIterable {
 /// - A256CBCHS512: [AES_256_CBC_HMAC_SHA_512](https://tools.ietf.org/html/rfc7518#section-5.2.5)
 public enum SymmetricKeyAlgorithm: String {
     case A256CBCHS512 = "A256CBC-HS512"
+    case A256CBCHS256 = "A256CBC-HS256"
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]:
```suggestion
    case A128CBCHS256 = "A128CBC-HS256"
```

See [RFC-7518](https://tools.ietf.org/html/rfc7518#section-5). `A128CBC-HS512` means AES in CBC mode with a 128 bit encryption key and HMAC with SHA-256 with the HMAC output truncated to 128 bits as per [Section 5.2.3](https://tools.ietf.org/html/rfc7518#section-5.2.3).

File: JOSESwift/Sources/Algorithms.swift [hunk 5]
```diff
@@ -91,13 +101,21 @@ public enum SymmetricKeyAlgorithm: String {
             }
 
             return (inputKey.subdata(in: 0..<32), inputKey.subdata(in: 32..<64))
+
+        case .A256CBCHS256:
+            guard checkKeyLength(for: inputKey) else {
+                throw JWEError.keyLengthNotSatisfied
+            }
+            return (inputKey.subdata(in: 0..<16), inputKey.subdata(in: 16..<32))
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: 👍

File: JOSESwift/Sources/Algorithms.swift [hunk 6]
```diff
@@ -91,13 +101,21 @@ public enum SymmetricKeyAlgorithm: String {
...
         }
     }
 
     func authenticationTag(for hmac: Data) -> Data {
         switch self {
         case .A256CBCHS512:
             return hmac.subdata(in: 0..<32)
+        case .A256CBCHS256:
+            return hmac.subdata(in: 0..<16)
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: 👍

File: JOSESwift/Sources/CryptoImplementation/AES.swift [hunk 7]
```diff
@@ -35,13 +35,19 @@ fileprivate extension SymmetricKeyAlgorithm {
         switch self {
         case .A256CBCHS512:
             return CCAlgorithm(kCCAlgorithmAES128)
+
+        case .A256CBCHS256:
+            return CCAlgorithm(kCCAlgorithmAES128)
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/CryptoImplementation/AES.swift]: 👍

File: JOSESwift/Sources/Algorithms.swift [hunk 8]
```diff
@@ -23,6 +23,7 @@
 //
 
 import Foundation
+import CommonCrypto
```
[the remark this answers is not in the document]
Comment (ramunasjurgilas) [reply JOSESwift/Sources/Algorithms.swift]: Done.

File: JOSESwift/Sources/Algorithms.swift (hunk 3 above)
Comment (ramunasjurgilas) [reply JOSESwift/Sources/Algorithms.swift]: Done

File: JOSESwift/Sources/Algorithms.swift (hunk 4 above)
Comment (ramunasjurgilas) [reply JOSESwift/Sources/Algorithms.swift]: Done.

**Comment by ramunasjurgilas** [comment]
Thanks for reviewing code.

Mentioned suggestion were applied. As well test case testRoundtrip_A128CBCHS256 was added. Do you @daniel-mohemian  think needs to add more test. 

Could you please review once again.

File: JOSESwift/Sources/Algorithms.swift [hunk 9]
```diff
@@ -54,32 +54,41 @@ public enum AsymmetricKeyAlgorithm: String, CaseIterable {
...
+            return 16
         }
     }
 
     func checkKeyLength(for key: Data) -> Bool {
         switch self {
         case .A256CBCHS512:
             return key.count == 64
+        case .A128CBCHS256:
+            return key.count == 32
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: 👍

File: JOSESwift/Sources/Algorithms.swift [hunk 10]
```diff
@@ -107,11 +124,14 @@ public enum SymmetricKeyAlgorithm: String {
...
 public enum HMACAlgorithm: String {
     case SHA512 = "SHA512"
+    case SHA256 = "SHA256"
 
     var outputLength: Int {
         switch self {
         case .SHA512:
             return 64
+        case .SHA256:
+            return 32
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: 👍

File: JOSESwift/Sources/CryptoImplementation/AES.swift [hunk 11]
```diff
@@ -35,13 +35,19 @@ fileprivate extension SymmetricKeyAlgorithm {
...
         }
     }
 
     func checkAESKeyLength(for key: Data) -> Bool {
         switch self {
         case .A256CBCHS512:
             return key.count == kCCKeySizeAES256
+
+        case .A128CBCHS256:
+            return key.count == kCCKeySizeAES128
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/CryptoImplementation/AES.swift]: :+1:

File: JOSESwift/Sources/CryptoImplementation/HMAC.swift [hunk 12]
```diff
@@ -29,6 +29,9 @@ fileprivate extension HMACAlgorithm {
         switch self {
         case .SHA512:
             return CCAlgorithm(kCCHmacAlgSHA512)
+            
+        case .SHA256:
+            return CCAlgorithm(kCCHmacAlgSHA256)
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/CryptoImplementation/HMAC.swift]: 👍

File: Tests/JWEDirectEncryptionTests.swift [hunk 13]
```diff
@@ -45,7 +45,22 @@ class JWEDirectEncryptionTests: RSACryptoTestCase {
     ])
 
     @available(*, deprecated)
-    func testRoundtrip() {
+
+    func testRoundtrip_A128CBCHS256() {
```
Comment (daniel-mohemian) [inline Tests/JWEDirectEncryptionTests.swift]:
```suggestion
    func testRoundtripA128CBCHS256() {
```

🐫

File: Tests/JWEDirectEncryptionTests.swift [hunk 14]
```diff
@@ -45,7 +45,22 @@ class JWEDirectEncryptionTests: RSACryptoTestCase {
...
+        let payload = Payload(data)
+        let encrypter = Encrypter(keyEncryptionAlgorithm: .direct, encryptionKey: symmetricKey, contentEncyptionAlgorithm: algorithm)!
+
+        let jwe = try! JWE(header: header, payload: payload, encrypter: encrypter)
+        let serialization = jwe.compactSerializedString
+
+        try! XCTAssertEqual(JWE(compactSerialization: serialization).decrypt(with: symmetricKey).data(), data)
+    }
+
+    func testRoundtrip_A256CBCHS512() {
```
Comment (daniel-mohemian) [inline Tests/JWEDirectEncryptionTests.swift]:
```suggestion
    func testRoundtrip_A256CBCHS512() {
```

🐫

File: JOSESwift/Sources/Algorithms.swift [hunk 15]
```diff
@@ -54,32 +54,41 @@ public enum AsymmetricKeyAlgorithm: String, CaseIterable {
 /// - A256CBCHS512: [AES_256_CBC_HMAC_SHA_512](https://tools.ietf.org/html/rfc7518#section-5.2.5)
 public enum SymmetricKeyAlgorithm: String {
     case A256CBCHS512 = "A256CBC-HS512"
+    case A128CBCHS256 = "A128CBC-HS256"
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: Please update the documentation comment with the new enum case.

File: JOSESwift/Sources/Algorithms.swift [hunk 16]
```diff
@@ -107,11 +124,14 @@ public enum SymmetricKeyAlgorithm: String {
 /// - SHA512
 public enum HMACAlgorithm: String {
     case SHA512 = "SHA512"
+    case SHA256 = "SHA256"
```
Comment (daniel-mohemian) [inline JOSESwift/Sources/Algorithms.swift]: Please update the documentation comment with the new enum case.

**Comment by ramunasjurgilas** [comment]
Thanks @daniel-mohemian for comments.

Additional test cases were added.
Very small documentation was added.

Could you please take a look once gain.

**Comment by ramunasjurgilas** [comment]
> 1 Warning
> ⚠️	Any changes to library code should be reflected in the changelog. Please consider adding a note in the _Unreleased_ section of the [CHANGELOG.md](https://github.com/airsidemobile/JOSESwift/blob/master/CHANGELOG.md).
> Generated by 🚫 [Danger](https://danger.systems/)

@daniel-mohemian Do I need to do some action on **Danger** comment?

File: Tests/AESEncrypterTests.swift [hunk 17]
```diff
@@ -35,8 +35,73 @@ class AESEncrypterTests: RSACryptoTestCase {
...
+        let hmacKey = keys.hmacKey
+        let encryptionKey = keys.encryptionKey
+
+        var concatData = String(data: additionalAuthenticatedData, encoding: .utf8)!.data(using: .utf8)!.base64URLEncodedData()
+        concatData.append(symmetricEncryptionContext.initializationVector)
+        concatData.append(symmetricEncryptionContext.ciphertext)
+        concatData.append(String(data: additionalAuthenticatedData, encoding: .utf8)!.data(using: .utf8)!.base64URLEncodedData().getByteLengthAsOctetHexData())
+
+        let keyLength = size_t(kCCKeySizeAES128)
+        var macOutData = Data(count: 64)
```
Comment (daniel-mohemian) [inline Tests/AESEncrypterTests.swift]:
``` suggestion
        var macOutData = Data(count: 32)
```

This is not too important as we only care about the first 16 bytes, but for correctness, this should be 32 (the output length in bytes of SHA256).

**Comment by ramunasjurgilas** [comment]
Suggested pull request change was added to the code.

**Comment by ramunasjurgilas** [comment]
I was glad, that I could help extending JOSESwift lib!

**Comment by daniel-mohemian** [comment]
Sorry for the delay @ramunasjurgilas. Your contribution is now live as [1.8.0](https://github.com/airsidemobile/JOSESwift/releases/tag/1.8.0). Thanks again! 🎉

**Pull request merged**